solutions/coci-16-zamjene.mdx

```diff
@@ -0,0 +1,269 @@
+
+If we assign each number a certain power of a prime $P$ ($42$ gets a power of $30$, $50$ gets a power of $35$, etc.),
+we can obtain our hash by multiplying each power by the number of times it occurs in the set.
+
+For example, if our component was the multiset ${3, 3, 4, 4, 5}$, and each number's power was itself, our hash would be:
+$$
+2 \cdot P^3 + 2 \cdot P^4 + 1 \cdot P^5
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a random value from a large range like $[1,10^6]$.
+
+</Info>
+
+Each component should not only have a hash of its number, but also a hash of the numbers it's supposed to have.
+For example, if our component was just the number $3$ but needs a $4$ to be sorted, its hash and required hash would be $P^3$ and $P^4$ respectively.
+Having this "required hash" allows us to quickly check if each component is sortable or not.
+
+When executing a union operation (aka a type 1 query), we can add the hashes and required hashes of the two components to get the hash of the combined component.
+When swapping two array elements (a type 2 query), we add and subtract a power of $P$ from each component, while leaving the required hashes untouched.
+
+### Tracking "Bad" Components
+
+However, when updating components, we also have to keep track of all the (henceforth referred to as "bad") components in the array to answer the output queries.
+
+First, we need the number of bad components after each modification.
+To answer a type 3 query, we check if there's any bad components.
+If there are, we output `NE`, and otherwise, we output `DA`.
+
+However, answering type 4 queries are a bit more tricky.
+We need some way to match up components such that the sum of their required hashes equals the sum of their actual hashes.
+In other words, if we let $x$ be the actual hash and $y$ be the required hash, we need to find all pairs of components ($i$ and $j$) such that the following is satisfied:
+$$
+x_i+x_j=y_i+y_j
+$$
+
+If we put the $i$s on one side, and $j$s on another, we get this:
+$$
+x_i-y_i=y_j-x_j
+$$
+
+This motivates us keeping a map of the difference between the required hash and actual hash and the number of positions whose components have such a difference.
+It doesn't matter whether we subtract the required hash from the actual hash or vice versa, as long as we stay consistent.
+We also keep track of the number of pairs of bad components that can be matched up.
+
+When adding or removing a bad component, we get its difference $d$ and see how many positions whose components have the difference $-d$ through the map to update the number of pairs.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(Q \cdot \alpha(N))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+class DominikArray {
```

Comment:
codesnip pls

Reply:
this class is the meat of the entire program

Reply:
oops im dumb ;w;